File: 43238.cpp
#include <string>
#include <vector>

#define MAX 100000000000000

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    long long from = 0;
    long long to = MAX;
    long long mid = (long long)((from + to) / 2);
    while(true) {
        if(from == to) return from;
        
        long long processed = 0;
        for(auto each : times) {
            processed += (long long)(mid / each);
        }
        if(processed >= n) {
            to = mid;
            mid = (long long)((from + to) / 2);
        } else {
            from = mid + 1;
            mid = (long long)((from + to) / 2);
        }
    }
    return answer;
}

File: 43238.swift
import Foundation

func solution(_ n:Int, _ times:[Int]) -> Int64 {
    return binarySearch(times, n, 0, 1000000000000000000)
}

func binarySearch(_ times: [Int], _ n: Int, _ from: Int64, _ to: Int64) -> Int64 {
    let mid = (from + to) / 2
    var current: Int64 = 0
    
    if from == to { return mid }
    
    for time in times {
        current += mid / Int64(time) 
    }
    
    if current < n {
        return binarySearch(times, n, mid + 1, to)
    } else {
        return binarySearch(times, n, from, mid)
    }
}
